libraries/lib-sample-track/MixerSource.cpp

```diff
@@ -123,30 +124,23 @@ size_t MixerSource::MixVariableRates(
          // Nothing to do if past end of play interval
          if (getLen > 0) {
             if (backwards) {
-               auto results =
-                  cache.GetFloats(*pos - (getLen - 1), getLen, mMayThrow);
-               if (results)
-                  memcpy(&queue[*queueLen], results, sizeof(float) * getLen);
-               else
-                  memset(&queue[*queueLen], 0, sizeof(float) * getLen);
-
+               // Not sure this works, haven't tested.
+               timePos -= getLen / track->GetRate();
```

Comment:
No.

This will conflict seriously with other refactoring I want you to review which introduces the wide sample sequence abstraction into the mixer.

And I repeat, we should seek a solution where cache maintenance is entirely internal to lib-wave-track, and control of cache lifetime is upstairs in Audacity, and it is all completely non-intrusive in lib-sample-track.

Reply:
> we should seek a solution where cache maintenance is entirely internal to lib-wave-track

Currently cache maintenance is in your to-be-introduced `lib-mixer`, through still persisting `SampleTrackCache`. So this proposal wouldn't be worse than what we have now (with respect to your concern).

Giving responsibility of cache maintenance to the consumers of `SampleTrack` (e.g. `MixerSource`, `SpectrogramCache`) makes the problems of eviction and cache miss avoidance easy. I cannot think of a solution that easy if this responsibility were to be that of `WaveTrack`.

Reply:
> This will conflict seriously with other refactoring I want you to review which introduces the wide sample sequence abstraction into the mixer.

I wouldn't mind so much rebasing my draft on top of your branch.

Reply:
> we should seek a solution where cache maintenance is entirely internal to lib-wave-track

If that's of concern, we could define an abstract wrapper for `SampleTrack` (to become `WideSampleSequence`), whose implementation (in `lib-wave-track` or a new lib only `lib-wave-track` would depend on) is injected in consumers of ´WaveTrack`.

Reply:
> > we should seek a solution where cache maintenance is entirely internal to lib-wave-track
> 
> Currently cache maintenance is in your to-be-introduced `lib-mixer`, through still persisting `SampleTrackCache`. So this proposal wouldn't be worse than what we have now (with respect to your concern).
> 
> Giving responsibility of cache maintenance to the consumers of `SampleTrack` (e.g. `MixerSource`, `SpectrogramCache`) makes the problems of eviction and cache miss avoidance easy. I cannot think of a solution that easy if this responsibility were to be that of `WaveTrack`.

The not-worse organization should become a positively better one.

Reply:
Maybe. If not I don't imagine this would get in the way of time stretching, meanwhile our current caching problem is.